node/src/components/rpc_server/rpcs.rs

```diff
@@ -6,69 +6,29 @@ pub mod account;
 pub mod chain;
 mod common;
 pub mod docs;
+mod error_code;
 pub mod info;
 pub mod state;
 
-use std::str;
+use std::{str, sync::Arc};
 
-use futures::{future::BoxFuture, TryFutureExt};
-use http::Response;
-use hyper::Body;
+use async_trait::async_trait;
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 use serde_json::Value;
-use warp::{
-    filters::BoxedFilter,
-    reject::{self, Reject},
-    Filter,
-};
-use warp_json_rpc::{filters, Builder};
 
+use casper_json_rpc::{Error, RequestHandlersBuilder, ReservedErrorCode};
 use casper_types::ProtocolVersion;
 
 use super::{ReactorEventT, RpcRequest};
 use crate::effect::EffectBuilder;
 pub use common::ErrorData;
 use docs::DocExample;
-
-/// The URL path.
-pub const RPC_API_PATH: &str = "rpc";
-
-/// Error code returned if the JSON-RPC response indicates failure.
-///
-/// See <https://www.jsonrpc.org/specification#error_object> for details.
-#[repr(i64)]
-enum ErrorCode {
-    NoSuchDeploy = -32000,
-    NoSuchBlock = -32001,
-    ParseQueryKey = -32002,
-    QueryFailed = -32003,
-    QueryFailedToExecute = -32004,
-    ParseGetBalanceURef = -32005,
-    GetBalanceFailed = -32006,
-    GetBalanceFailedToExecute = -32007,
-    InvalidDeploy = -32008,
-    NoSuchAccount = -32009,
-    FailedToGetDictionaryURef = -32010,
-    FailedToGetTrie = -32011,
-    NoSuchStateRoot = -32012,
-    // Same error code as warp_json INTERNAL_ERROR.
-    InternalError = -32063,
-}
-
-#[derive(Debug)]
-pub(super) struct Error(String);
-
-impl Reject for Error {}
-
-impl From<anyhow::Error> for Error {
-    fn from(error: anyhow::Error) -> Self {
-        Error(error.to_string())
-    }
-}
+pub use error_code::ErrorCode;
 
 /// A JSON-RPC requiring the "params" field to be present.
-pub trait RpcWithParams {
+#[async_trait]
+pub(super) trait RpcWithParams {
```

Comment:
Maybe we can move that to `casper-json-rpc`? The trait seems rather generic and is specific to JSON RPC server.

Reply:
Yes, I toyed with that too.  It would be nice since that's about the only place users have to worry about returning a `ReservedErrorCode` (other than if they want to return an internal server error), but I decided against that since I think we'd need to split the current trait up a bit as it has the node-specific `EffectBuilder<REv>` in some of the methods.

We could definitely look to do that in the future, but I did want to try and keep the changes to the existing node codebase as limited as possible in this PR.